Add collapse in hybrid query documentation (#10073)
opensearch-project/documentation-website: _vector-search/ai-search/hybrid-search/collapse.md

---
layout: default
title: Using collapse with a hybrid query
parent: Hybrid search
grand_parent: AI search
has_children: false
nav_order: 10
---

# Using collapse with hybrid query
**Introduced 3.1**
{: .label .label-purple }

You can collapse on a field by specifying the `collapse` parameter in the search query.
This will return the highest scoring document for each unique value of that field, up to the specified size in the query.

> The collapse parameter requires the field being collapsed to be of either a `keyword` or a `numeric` type.

Reviewer: please add what will happen if field is neither keyword nor numeric, and if that's the same for other query types (I assume we follow general pattern for supported field types)

> Note: Inner hits compatibility within the collapse description is not currently supported for hybrid query.

Reviewer: I think we need to add more notes, not sure how to organize them though:
- we don't support collapse via processor (I think we do not), maybe more details here: will there be an error, or results will be off etc.
- performance impact may be higher with large result sets
- aggregations run on pre-collapsed results 
- collapse affects pagination and total hit counts

Reviewer: Let's add middle two and the pagination.
As for the processor, I think results can be different from one that we can see when we pass a parameter, due to processor not seeing all search hits from shards. It's same for other queries, but I think it's worth mentioning.

The following example is taken from the documentation for [collapse in non-hybrid query](https://docs.opensearch.org/docs/latest/search-plugins/collapse-search/) and modified for hybrid query:

Create an index:
```json
PUT /bakery-items
{
  "mappings": {
    "properties": {
      "item": {
        "type": "keyword"
      },
      "category": {
        "type": "keyword"
      },
      "price": {
        "type": "float"
      },
      "baked_date": {
        "type": "date"
      }
    }
  }
}
```

Index documents:
```json
POST /bakery-items/_bulk
{ "index": {} }
{ "item": "Chocolate Cake", "category": "cakes", "price": 15, "baked_date": "2023-07-01T00:00:00Z" }
{ "index": {} }
{ "item": "Chocolate Cake", "category": "cakes", "price": 18, "baked_date": "2023-07-04T00:00:00Z" }
{ "index": {} }

> { "item": "Vanilla Cake", "category": "cakes", "price": 12, "baked_date": "2023-07-02T00:00:00Z" }

Reviewer 2: Can you add 2-3 items of `Vanilla Cake` and then show it got collapsed to just 1 the response?

              "match": {
                "category": "cakes"
              }
            }
          }
        }
      ]
    }
  },
  "collapse": {
    "field": "item"
  }
}
```

Response:
```json
{
  "took": 2,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 3,
      "relation": "eq"
    },
    "max_score": 1.0,
    "hits": [
      {
        "_index": "bakery-items",
        "_id": "KS-5YJcBeeOHd7EbgdPT",
        "_score": 1.0,
        "_source": {
          "item": "Chocolate Cake",
          "category": "cakes",
          "price": 15,
          "baked_date": "2023-07-01T00:00:00Z"
        },
        "fields": {
          "item": [
            "Chocolate Cake"
          ]
        }
      },
      {
        "_index": "bakery-items",
        "_id": "Ky-5YJcBeeOHd7EbgdPT",
        "_score": 0.5005,
        "_source": {
          "item": "Vanilla Cake",
          "category": "cakes",
          "price": 12,
          "baked_date": "2023-07-02T00:00:00Z"
        },
        "fields": {
          "item": [
            "Vanilla Cake"
          ]
        }
      }
    ]
  }
}
```

> You can also use sort in addition to collapsing the documents:

Reviewer: this section needs more context: why only sort, how about other features like pagination or explain etc. Is this combination of sort and collapse work differently comparing to other queries?

Author: No, it works the exact same.  Should I just remove this section?

---
layout: default
title: Using collapse with a hybrid query
parent: Hybrid search
grand_parent: AI search
has_children: false
nav_order: 10
---

# Using collapse with hybrid query
**Introduced 3.1**
{: .label .label-purple }

You can collapse on a field by specifying the `collapse` parameter in the search query.
This will return the highest scoring document for each unique value of that field, up to the specified size in the query.
The collapse parameter requires the field being collapsed to be of either a `keyword` or a `numeric` type.

There are a few important considerations to keep in mind when using collapse with hybrid query.
* Inner hits compatibility within the collapse description is not currently supported for hybrid query.
* Performance impact may be higher when using large result sets.
* Aggregations run on pre-collapsed results.

> * Collapse affects the paginated results.

Author: We can rephrase, but search_after is not supported with collapse (for all query types)

              "match": {
                "category": "cakes"
              }
            }
          }
        }
      ]
    }
  },
  "collapse": {
    "field": "item"
  }
}
```

Response:
```json
{
  "took": 5,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 5,
      "relation": "eq"
    },
    "max_score": 1.0,
    "hits": [
      {
        "_index": "bakery-items",
        "_id": "wBRPZZcB49c_2-1rYmO7",
        "_score": 1.0,
        "_source": {
          "item": "Chocolate Cake",
          "category": "cakes",
          "price": 15,
          "baked_date": "2023-07-01T00:00:00Z"
        },
        "fields": {
          "item": [
            "Chocolate Cake"
          ]
        }
      },
      {
        "_index": "bakery-items",
        "_id": "whRPZZcB49c_2-1rYmO7",
        "_score": 0.5005,
        "_source": {
          "item": "Vanilla Cake",
          "category": "cakes",
          "price": 12,
          "baked_date": "2023-07-02T00:00:00Z"
        },
        "fields": {
          "item": [
            "Vanilla Cake"
          ]
        }
      }
    ]
  }
}
```

> Collapse is compatible with other features in hybrid query using syntax that is standard for those features, such as sort, pagination, and explain.

Author: The size means that only results from the top `size` groups are passed to the normalization workflow

> Collapse is compatible with other features in hybrid query using syntax that is standard for those features, such as sort, pagination, and explain.

Author: Maybe something like this from @martin-gaievski 's comment above:
```
Collapse affects pagination by reducing result counts per page, altering result distribution across pages.
```

---
layout: default
title: Using collapse with a hybrid query
parent: Hybrid search
grand_parent: AI search
has_children: false
nav_order: 10
---

# Using collapse with hybrid query
**Introduced 3.1**
{: .label .label-purple }

You can collapse on a field by specifying the `collapse` parameter in the search query.
This will return the highest scoring document for each unique value of that field, up to the specified size in the query.
The collapse parameter requires the field being collapsed to be of either a `keyword` or a `numeric` type.

There are a few important considerations to keep in mind when using collapse with hybrid query.
* Inner hits compatibility within the collapse description is not currently supported for hybrid query.
* Performance impact may be higher when using large result sets.
* Aggregations run on pre-collapsed results.

> * Collapse affects the pagination by reducing the number of results and altering the result distribution across pages.  Consider setting a higher pagination depth to retrieve more results.

Reviewer 2: ```suggestion
* Collapse affects the pagination by reducing the number of results and altering the result distribution across pages. Consider setting a higher pagination depth to retrieve more results.
```

        },
        "fields": {
          "item": [
            "Chocolate Cake"
          ]
        }
      },
      {
        "_index": "bakery-items",
        "_id": "whRPZZcB49c_2-1rYmO7",
        "_score": 0.5005,
        "_source": {
          "item": "Vanilla Cake",
          "category": "cakes",
          "price": 12,
          "baked_date": "2023-07-02T00:00:00Z"
        },
        "fields": {
          "item": [
            "Vanilla Cake"
          ]
        }
      }
    ]
  }
}
```

Collapse is compatible with other features in hybrid query using syntax that is standard for those features, such as sort, pagination, and explain.
For example, this is how you can combine collapse and sort:

```json
GET /bakery-items/_search?search_pipeline=norm-pipeline
{
  "query": {
    "hybrid": {
      "queries": [
        {
          "match": {
                "item": "Chocolate Cake"
          }
        },
        {
          "bool": {
                "must": {
                    "match": {
                        "category": "cakes"
                    }
                }
          }
        }
      ]
    }
  },
  "collapse": {
    "field": "item"
  },

>   "sort": "price"

Reviewer 2: Nit: You can add order_by within the sort clause

Author: This is the same as the example in the main collapse documentation, I think we should keep it for the simplicity

}
```

Response:
```json
{
  "took": 13,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 5,
      "relation": "eq"
    },
    "max_score": null,
    "hits": [
      {
        "_index": "bakery-items",
        "_id": "whRPZZcB49c_2-1rYmO7",
        "_score": null,
        "_source": {
          "item": "Vanilla Cake",
          "category": "cakes",
          "price": 12,
          "baked_date": "2023-07-02T00:00:00Z"
        },
        "fields": {
          "item": [
            "Vanilla Cake"
          ]
        },
        "sort": [
          12.0
        ]
      },
      {
        "_index": "bakery-items",
        "_id": "wBRPZZcB49c_2-1rYmO7",
        "_score": null,
        "_source": {
          "item": "Chocolate Cake",
          "category": "cakes",
          "price": 15,
          "baked_date": "2023-07-01T00:00:00Z"
        },
        "fields": {
          "item": [
            "Chocolate Cake"
          ]
        },
        "sort": [
          15.0
        ]
      }
    ]
  }
}
```

> This is how you can combine collapse and pagination:

Author: It's more that the size is one so it's picking up the second document instead of the first

                "item": "Chocolate Cake"
          }
        },
        {
          "bool": {
                "must": {
                    "match": {
                        "category": "cakes"
                    }
                }
          }
        }
      ]
    }
  },
  "collapse": {
    "field": "item"
  },
  "from": 1,
  "size": 1
}
```

Response:
```json
{
    "took": 4,
    "timed_out": false,
    "_shards": {
        "total": 1,
        "successful": 1,
        "skipped": 0,
        "failed": 0
    },
    "hits": {
        "total": {
            "value": 5,
            "relation": "eq"
        },
        "max_score": 1.0,
        "hits": [
            {
                "_index": "bakery-items",
                "_id": "35ILepcBheX09_dPt8TD",
                "_score": 0.5005,
                "_source": {
                    "item": "Vanilla Cake",
                    "category": "cakes",
                    "price": 12,
                    "baked_date": "2023-07-02T00:00:00Z"
                },
                "fields": {
                    "item": [
                        "Vanilla Cake"
                    ]
                }
            }
        ]
    }
}
```

> This is how you can combine collapse and explain:

Author: Ah I must have accidentally deleted that

The `collapse` parameter is compatible with other hybrid query search options, such as sort, explain, and pagination, using their standard syntax.

When using `collapse` in a hybrid query, note the following considerations:

- Inner hits are not supported.
- Performance may be impacted when working with large result sets.
- Aggregations run on pre-collapsed results, not the final output.
- Pagination behavior changes: Because `collapse` reduces the total number of results, it can affect how results are distributed across pages. To retrieve more results, consider increasing the pagination depth.
- Results may differ from those returned by the [`collapse` response processor]({{site.url}}{{site.baseurl}}/search-plugins/search-pipelines/collapse-processor/), which applies collapse logic after the query is executed.

## Example

> The following example demonstrates collapsing hybrid query results.

Reviewer 3: ```suggestion
The following example demonstrates how to collapse hybrid query results.
```

          "baked_date": "2023-07-01T00:00:00Z"
        },
        "fields": {
          "item": [
            "Chocolate Cake"
          ]
        }
      },
      {
        "_index": "bakery-items",
        "_id": "whRPZZcB49c_2-1rYmO7",
        "_score": 0.5005,
        "_source": {
          "item": "Vanilla Cake",
          "category": "cakes",
          "price": 12,
          "baked_date": "2023-07-02T00:00:00Z"
        },
        "fields": {
          "item": [
            "Vanilla Cake"
          ]
        }
      }
    ]
  }
```

## Collapse and sort results

To collapse and sort hybrid query results, provide the `collapse` and `sort` parameters in the query:

```json
GET /bakery-items/_search?search_pipeline=norm-pipeline
{
  "query": {
    "hybrid": {
      "queries": [
        {
          "match": {
                "item": "Chocolate Cake"
          }
        },
        {
          "bool": {
                "must": {
                    "match": {
                        "category": "cakes"
                    }
                }
          }
        }
      ]
    }
  },
  "collapse": {
    "field": "item"
  },
  "sort": "price"
}
```
{% include copy-curl.html %}

> For more information about sorting in hybrid query, see [Using sorting with a hybrid query]({{site.url}}{{site.baseurl}}/vector-search/ai-search/hybrid-search/sorting/).

Reviewer 3: ```suggestion
For more information about sorting in a hybrid query, see [Using sorting with a hybrid query]({{site.url}}{{site.baseurl}}/vector-search/ai-search/hybrid-search/sorting/).
```

                    "category": "cakes",
                    "price": 12,
                    "baked_date": "2023-07-02T00:00:00Z"
                },
                "fields": {
                    "item": [
                        "Vanilla Cake"
                    ]
                },
                "_explanation": {
                    "value": 1.0,
                    "description": "combined score of:",
                    "details": [
                        {
                            "value": 0.0,
                            "description": "ConstantScore(item:Chocolate Cake) doesn't match id 2",
                            "details": []
                        },
                        {
                            "value": 1.0,
                            "description": "ConstantScore(category:cakes)",
                            "details": []
                        }
                    ]
                }
            }
        ]
    }
```

For more information about using `explain` in a hybrid query, see [Hybrid search explain]({{site.url}}{{site.baseurl}}/vector-search/ai-search/hybrid-search/explain/).

## Collapse and pagination

> You can paginate collapsed results by providing the `from` and `size` parameters. For more information about pagination in a hybrid query, see [Paginating hybrid query results]({{site.url}}{{site.baseurl}}/vector-search/ai-search/hybrid-search/pagination/). For more information about `from` and `size`, see 

Reviewer 3: The last sentence is incomplete.

{ "index": {} }
{ "item": "Carrot Cake", "category": "cakes", "price": 26, "baked_date": "2023-07-21T00:00:00Z" }
{ "index": {} }
{ "item": "Red Velvet Cake", "category": "cakes", "price": 25, "baked_date": "2023-07-09T00:00:00Z" }
{ "index": {} }
{ "item": "Red Velvet Cake", "category": "cakes", "price": 29, "baked_date": "2023-07-30T00:00:00Z" }
{ "index": {} }
{ "item": "Cheesecake", "category": "cakes", "price": 27. "baked_date": "2023-07-09T00:00:00Z" }
{ "index": {} }
{ "item": "Cheesecake", "category": "cakes", "price": 34. "baked_date": "2023-07-21T00:00:00Z" }
{ "index": {} }
{ "item": "Coffee Cake", "category": "cakes", "price": 42, "baked_date": "2023-07-09T00:00:00Z" }
{ "index": {} }
{ "item": "Coffee Cake", "category": "cakes", "price": 41, "baked_date": "2023-07-05T00:00:00Z" }
{ "index": {} }
{ "item": "Cocunut Cake", "category": "cakes", "price": 23, "baked_date": "2023-07-09T00:00:00Z" }
{ "index": {} }
{ "item": "Cocunut Cake", "category": "cakes", "price": 32, "baked_date": "2023-07-12T00:00:00Z" }
// Additional documents omitted for brevity
```
{% include copy-curl.html %}

> Run a `hybrid` query, specifying the `from` and `size` parameters to paginate results. In the following example, the query requests two results starting from the sixth position (`from: 5, size: 2`). The pagination depth is set to limit each shard to return a maximum of 10 documents. After the results are retrieved, the `collapse` parameter is applied to group them by the `item` field:

Reviewer 3: ```suggestion
Run a `hybrid` query, specifying the `from` and `size` parameters to paginate results. In the following example, the query requests two results starting from the sixth position (`from: 5, size: 2`). The pagination depth is set to limit each shard to return a maximum of 10 documents. After the results are retrieved, the `collapse` parameter is applied in order to group them by the `item` field:
```
